src/Microsoft.AspNetCore.ResponseCaching/Internal/DefaultResponseCacheEntrySerializer.cs

```diff
@@ -154,7 +160,18 @@ private static void WriteCachedResponse(BinaryWriter writer, CachedResponse entr
         private static void WriteCachedVaryBy(BinaryWriter writer, CachedVaryBy entry)
         {
             writer.Write(nameof(CachedVaryBy));
-            writer.Write(entry.Headers);
+
+            writer.Write(entry.Headers.Count);
+            if (entry.Headers.Count > 0)
+            {
+                writer.Write(entry.Headers);
```

Comment:
You can save yourself some joins and splits by adding a for loop here and writing out the headers and param values separately.